src/Uno.UI/UI/Xaml/Shapes/Shape.layout.cs

```diff
@@ -277,53 +274,75 @@ private protected Size MeasureAbsoluteShape(Size availableSize, NativePath? path
 					// Also, as the path does not have any notion of stroke thickness, we have to include it for the measure phase.
 					// Note: The logic would say to include the full StrokeThickness as it will "overflow" half on booth side of the path,
 					//		 but WinUI does include only the half of it.
-					var pathNaturalSize = new Size(
-						pathBounds.X == 0 ? pathBounds.Width + strokeThickness : pathBounds.Right + alignedHalfStrokeThickness,
-						pathBounds.Y == 0 ? pathBounds.Height + strokeThickness : pathBounds.Bottom + alignedHalfStrokeThickness);
-					size = pathNaturalSize.AtMost(userMaxSize).AtLeast(userMinSize); // The size defined on the Shape has priority over the size of the geometry itself!
+					var halfStrokeThickness = strokeThickness / 2;
+					size = new Size(pathBounds.Right + halfStrokeThickness, pathBounds.Bottom + halfStrokeThickness);
 					break;
 
 				case Stretch.Fill:
-					size = userMaxSize.FiniteOrDefault(availableSize.FiniteOrDefault(pathSize));
+					size = new Size(
+						availableSize.Width == PositiveInfinity ? pathBounds.Width + strokeThickness : Math.Max(availableSize.Width, strokeThickness),
+						availableSize.Height == PositiveInfinity ? pathBounds.Height + strokeThickness : Math.Max(availableSize.Height, strokeThickness)
+						);
 					break;
 
-#if !IS_DESIRED_SMALLER_THAN_CONSTRAINTS_ALLOWED
-				case Stretch.Uniform when (userSize.min.hasWidth && userSize.min.width > availableSize.Width) || (userSize.min.hasHeight && userSize.min.height > availableSize.Height):
-					size = availableSize;
-					break;
-
-				// Note: If the parent is going to stretch us due to the Width and/or the Height, we still go in the case below,
-				//		 so we compute the effective size and we properly full-fill the _realDesiredSize
-#endif
-
 				case Stretch.Uniform:
-					size = userMaxSize.FiniteOrDefault(availableSize);
-					renderScale = ComputeScaleFactors(pathSize, ref size, strokeThickness);
-					if (renderScale.x > renderScale.y)
+					// For Stretch.Uniform, if available size is infinity (both Width and Height), we just return the path size plus stroke thickness.
+					if (availableSize.Width == PositiveInfinity && availableSize.Height == PositiveInfinity)
```

Comment:
Personally, I would merge the 3 cases here directly in the `switch` (i.e. `case Stretch.Uniform when availableSize.Width == PositiveInfinity && availableSize.Height == PositiveInfinity`),  would be cutter I think ... but personal preferences (same for the `UniformToFill`).

And about the changes you made on the computations ... I would assume they are fine, let see if the CI is fine too :)

Reply:
> I would assume they are fine, let see if the CI is fine too :)

It seems they are fine. I went through many samples, and all the differences I could see were differences from WinUI that are now fixed. I also added examples of "Before", "After", and "WinUI" screenshots in PR description.